src/state/write_batch.rs

```diff
@@ -0,0 +1,441 @@
+/*
+    Copyright © 2023, ParallelChain Lab
+    Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
+*/
+//! This module defines the [`WriteBatch`] trait for batching writes to the key-value store, and its
+//! implementation for the [block tree](crate::state::block_tree::BlockTree), the [`BlockTreeWriteBatch`].
+pub struct BlockTreeWriteBatch<W: WriteBatch>(pub(super) W);
+
+use borsh::BorshSerialize;
+use paths::*;
+
+use crate::hotstuff::types::QuorumCertificate;
+use crate::pacemaker::types::TimeoutCertificate;
+use crate::types::basic::{
+    AppStateUpdates, BlockHeight, ChildrenList, CryptoHash, DataLen, ViewNumber,
+};
+use crate::types::block::Block;
+use crate::types::validators::{
+    ValidatorSet, ValidatorSetBytes, ValidatorSetUpdates, ValidatorSetUpdatesStatusBytes,
+};
+
+use super::block_tree::BlockTreeError;
+use super::kv_store::Key;
+use super::paths;
+use super::utilities::combine;
+
+pub trait WriteBatch {
+    fn new() -> Self;
+    fn set(&mut self, key: &[u8], value: &[u8]);
+    fn delete(&mut self, key: &[u8]);
+}
+
+impl<W: WriteBatch> BlockTreeWriteBatch<W> {
+    pub(crate) fn new() -> BlockTreeWriteBatch<W> {
+        BlockTreeWriteBatch(W::new())
+    }
+
+    pub fn new_unsafe() -> BlockTreeWriteBatch<W> {
+        Self::new()
+    }
+
+    /* ↓↓↓ Block ↓↓↓  */
+
+    pub fn set_block(&mut self, block: &Block) -> Result<(), BlockTreeError> {
+        let block_prefix = combine(&paths::BLOCKS, &block.hash.bytes());
+
+        self.0.set(
+            &combine(&block_prefix, &paths::BLOCK_HEIGHT),
+            &block
+                .height
+                .try_to_vec()
+                .map_err(|err| KVSetError::SerializeValueError {
+                    key: Key::BlockHeight {
+                        block: block.hash.clone(),
+                    },
+                    source: err,
+                })?,
+        );
+        self.0.set(
+            &combine(&block_prefix, &paths::BLOCK_JUSTIFY),
+            &block
+                .justify
+                .try_to_vec()
+                .map_err(|err| KVSetError::SerializeValueError {
+                    key: Key::BlockJustify {
+                        block: block.hash.clone(),
+                    },
+                    source: err,
+                })?,
+        );
+        self.0.set(
+            &combine(&block_prefix, &paths::BLOCK_DATA_HASH),
+            &block
+                .data_hash
+                .try_to_vec()
+                .map_err(|err| KVSetError::SerializeValueError {
+                    key: Key::BlockDataHash {
+                        block: block.hash.clone(),
+                    },
+                    source: err,
+                })?,
+        );
+        self.0.set(
+            &combine(&block_prefix, &paths::BLOCK_DATA_LEN),
+            &block
+                .data
+                .len()
```

Comment:
Bug: `.len()` here returns a `u64`, which is then serialized and put into the world state. But on the reading end, it is deserialized into a `DataLen`, which is a newtype around `u32`. This does not cause an error or a panic because if `BorshDeserialize::deserialize` is called with a buffer that is longer than expected, it simply points the start of the buffer to the start of the remaining byte, but leads to incorrect executions.

Reply:
Will fix with https://github.com/parallelchain-io/hotstuff_rs/issues/44